kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -66,245 +82,410 @@ String filterDescriptor() {
     }
 
     @Override
-    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
-        if (msg instanceof InternalRequestFrame<?>) {
-            // jump the queue, internal request must flow!
-            doWrite(ctx, msg, promise);
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.channelActive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof InternalResponseFrame<?> decodedFrame) {
+            // jump the queue, let responses to asynchronous requests flow back to their sender
+            if (decodedFrame.isRecipient(filter)) {
+                completeInternalResponse(decodedFrame);
+            }
+            else {
+                readDecodedResponse(decodedFrame);
+            }
         }
-        else if (writeFuture.isDone()) {
-            writeFuture = doWrite(ctx, msg, promise);
+        else if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
+            if (readFuture.isDone()) {
+                readFuture = readDecodedResponse(decodedFrame);
+            }
+            else {
+                readFuture = readFuture.whenComplete((a, b) -> {
```

Comment:
I think I've found a bug in this and the equivalent write code, it doesn't queue properly. If you have multiple actions in the queue, the first future completion will call `readDecodedResponse` on all of the enqueued actions. This is visible when you set up the test with two deferred futures followed by a nondeferred future, the response is out of order because completing future1 triggers both actions.

Here's a fix, switching to thenCompose to chain the futures https://github.com/k-wall/kroxylicious/pull/7

Also I was wondering if we need an immediate `flush` after each deferred write finishes, if we wait for the whenComplete on the `writeFuture` then we may be delayed while other async work completes, where we were able to flush it to the broker straight away.

Reply:
Existing bug so we could leave it out of this PR

Reply:
The change looks good.  I'll keep it in this PR, but merged as separate commit.